opentelemetry-api/src/opentelemetry/metrics/instrument.py

```diff
@@ -0,0 +1,204 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: disable=too-many-ancestors
+
+# type: ignore
+
+
+from abc import ABC, abstractmethod
+from logging import getLogger
+from re import compile as compile_
+from threading import RLock
+from types import GeneratorType
+
+from opentelemetry.metrics.measurement import Measurement
+
+_logger = getLogger(__name__)
+
+
+class Instrument(ABC):
+
+    _name_regex = compile_(r"[a-zA-Z][-.\w]{0,62}")
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def unit(self):
+        return self._unit
+
+    @property
+    def description(self):
+        return self._description
+
+    @abstractmethod
+    def __init__(self, name, *args, unit="", description="", **kwargs):
+        self._lock = RLock()
+
+        if name is None or self._name_regex.fullmatch(name) is None:
+            _logger.error("Invalid instrument name %s", name)
+
+        else:
+            self._name = name
+
+        if unit is None:
+            self._unit = ""
+        elif len(unit) > 63:
+            _logger.error("unit must be 63 characters or shorter")
+
+        elif any(ord(character) > 127 for character in unit):
+            _logger.error("unit must only contain ASCII characters")
+        else:
+            self._unit = unit
+
+        if description is None:
+            description = ""
+
+        self._description = description
+
+
+class Synchronous(Instrument):
+    pass
+
+
+class Asynchronous(Instrument):
+    @abstractmethod
+    def __init__(
+        self, name, callback, *args, unit="", description="", **kwargs
+    ):
+        super().__init__(name, *args, unit=unit, description="", **kwargs)
+        with self._lock:
```

Comment:
I don't think we need any locking in the API instruments since they don't modify any shared memory

Reply:
I don't disagree, but it is required [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#concurrency).

Reply:
The code is thread safe without the lock.

Reply:
It's not required by the statement.   THis method needs to be thread-safe, but thread-safe != locks.

As long as the aggregator you send this to handles things concurrently, you should be ok.   E.g. in Java we're using a lock-less (compare-and-swap atomics) for as much of this as we can.

Reply:
I have removed all the locks in this PR, I see it now as a relatively minor implementation detail and it is not worth it to stop this PR just for this.